### PR TITLE
Self consumption predictor

### PR DESCRIPTION
**Pull request by Lasall**
* Inverter: Self consumption interpolator for better discharge_hour results
 * Small penalty when EV 100% and charge >0
 * Price Forceast (use mean of last 7 days instead of repeat)
 * Price Prediction as JSON simulation output, config fixed electricty fees configurable + MyPy & Ruff

- [x] provide /strompreis 7 days mean (datasequence? mean). Following PRs are not fully included in this PR: #268 #284 #285 (regarding price_forecast.py and /strompreis endpoint) -> done in #281

This is the merge of the recent PRs from main into our current development branch

### CODE REVIEW

File: src/akkudoktoreos/server/fastapi_server.py [hunk 1]
```diff
@@ -106,6 +106,25 @@ def fastapi_strompreis() -> list[float]:
     specific_date_prices = marketprice_series.loc[
         prediction_eos.start_datetime : prediction_eos.end_datetime
     ]
+    # TODO merge
```
Comment (Lasall) [inline src/akkudoktoreos/server/fastapi_server.py:109]: @b0661 I need help here from merging the latest changes from main. I don't know how to proceed here (add charges to electricity provider, provide mean, ...) See changes to src/akkudoktoreos/prediction/price_forecast.py

File: src/akkudoktoreos/prediction/self_consumption_probability.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
+#!/usr/bin/env python
```
Comment (b0661) [inline src/akkudoktoreos/prediction/self_consumption_probability.py:1]: The use of underscores in Python module names is discouraged (see https://peps.python.org/pep-0008/#package-and-module-names). Maybe you can find a shorter module name with less/no underscores. Just nitpicking.

File: src/akkudoktoreos/server/fastapi_server.py (hunk 1 above)
Comment (b0661) [reply src/akkudoktoreos/server/fastapi_server.py:109]: Price forecast using akkudoktor.net is in src/akkudoktoreos/prediction/elecpriceakkudoktor.py. The adaptions have to be made here:

```Python
        # Assure price ends at end_time
        if compare_datetimes(self[-1].date_time, self.end_datetime).lt:
            record = ElecPriceDataRecord(
                date_time=self.end_datetime,
                elecprice_marketprice=self[-1].elecprice_marketprice,
            )
            self.append(record)
```

As akkudoktor.net currently only provides 24 prices in 1 hour intervals. Currently the last price available from akkudoktor.net is set also on 'self.end_datetime'. When retrieving a price array the missing prices in between will be interpolated using the `key_to_array()` function.

To add Andreas' changes you have to add the missing prices here (see above).  I think he made some calculations to get the missing 24 prices using the data retrieved from akkudoktor.net. Currently not enough past data is retrieved from akkudoktor net and only `self.config.prediction_historic_hours` (typically 24 hours) past data is stored in the prediction data records.

```Python
    @cache_in_file(with_ttl="1 hour")
    def _request_forecast(self) -> AkkudoktorElecPrice:
        """Fetch electricity price forecast data from Akkudoktor API.

        This method sends a request to Akkudoktor's API to retrieve forecast data for a specified
        date range. The response data is parsed and returned as JSON for further processing.

        Returns:
            dict: The parsed JSON response from Akkudoktor API containing forecast data.

        Raises:
            ValueError: If the API response does not include expected `electricity price` data.
        """
        source = "https://api.akkudoktor.net"
        date = to_datetime(self.start_datetime, as_string="Y-M-D")
        last_date = to_datetime(self.end_datetime, as_string="Y-M-D")
        response = requests.get(
            f"{source}/prices?date={date}&last_date={last_date}&tz={self.config.timezone}"
        )
        response.raise_for_status()  # Raise an error for bad responses
        logger.debug(f"Response from {source}: {response}")
        akkudoktor_data = self._validate_data(response.content)
        # We are working on fresh data (no cache), report update time
        self.update_datetime = to_datetime(in_timezone=self.config.timezone)
        return akkudoktor_data
```

I assume `date` has to be changed to be ??7?? days before `self.start_datetime`. Then you get all the data to do the extra calculations. The data will be in `akkudoktor_data`, which you can use to do the extra calculations.

```Python
    from akkudoktoreos.utils.datetimeutils import to_duration
    date = to_datetime(self.start_datetime  - to_duration("7 days"), as_string="Y-M-D")
```

**Comment by drbacke** [comment]
Please be carfull here. There can be a recursive / infininte loop between get_price_for_daterange and get_7dmean_price() very fast.

File: src/akkudoktoreos/server/fastapi_server.py (hunk 1 above)
Comment (b0661) [reply src/akkudoktoreos/server/fastapi_server.py:109]: Will add solution to next config/ prediction update/fix.

File: src/akkudoktoreos/core/ems.py [hunk 3]
```diff
@@ -265,151 +271,164 @@ def set_start_hour(self, start_hour: Optional[int] = None) -> None:
...
 
         akku_soc_pro_stunde begin of the hour, initial hour state!
         last_wh_pro_stunde integral of  last hour (end state)
         """
         # Check for simulation integrity
         if (
-            self.gesamtlast is None
-            or self.pv_prognose_wh is None
-            or self.strompreis_euro_pro_wh is None
+            self.load_energy_array is None
```
Comment (NormannK) [inline src/akkudoktoreos/core/ems.py:286]: more than one can be missing and all should be reported:
```
    missing_data = []

    if self.load_energy_array is None:
        missing_data.append("Load Curve")
    if self.pv_prediction_wh is None:
        missing_data.append("PV Forecast")
    if self.elect_price_hourly is None:
        missing_data.append("Electricity Price")
    if self.ev_charge_hours is None:
        missing_data.append("EV Charge Hours")
    if self.ac_charge_hours is None:
        missing_data.append("AC Charge Hours")
    if self.dc_charge_hours is None:
        missing_data.append("DC Charge Hours")
    if self.elect_revenue_per_hour_arr is None:
        missing_data.append("Feed-in Tariff")

    if missing_data:
        error_msg = (
            "Mandatory data missing - "
            + ", ".join(missing_data)
        )
        logger.error(error_msg)
        raise ValueError(error_msg)
```

Comment (Lasall) [reply src/akkudoktoreos/core/ems.py:286]: this is too complicated for mypy :(

Comment (NormannK) [reply src/akkudoktoreos/core/ems.py]: lol ok something for later

File: src/akkudoktoreos/devices/devices.py [hunk 4]
```diff
@@ -159,16 +160,20 @@ def total_losses_wh(self) -> float:
     # Devices
     # TODO: Make devices class a container of device simulation providers.
     #       Device simulations to be used are then enabled in the configuration.
-    akku: ClassVar[Battery] = Battery(provider_id="GenericBattery")
-    eauto: ClassVar[Battery] = Battery(provider_id="GenericBEV")
+    battery: ClassVar[Battery] = Battery(provider_id="GenericBattery")
+    ev: ClassVar[Battery] = Battery(provider_id="GenericBEV")
```
Comment (NormannK) [inline src/akkudoktoreos/devices/devices.py:164]: TODO: rename BEV to ev

**Review by NormannK — APPROVED** [review]
code is good
i dont understand the Self consumption predictor

**Pull request merged**